SwiftCompilerSources/Sources/Optimizer/FunctionPasses/AutodiffClosureSpecialization.swift

```diff
@@ -0,0 +1,710 @@
+                                                 convertedAndReabstractedClosures: &convertedAndReabstractedClosures)
+  }
+
+  if let cvt = convertedAndReabstractedClosure as? ConvertFunctionInst {
+    convertedAndReabstractedClosures.insert(cvt)
+    return 
+      markConvertedAndReabstractedClosuresAsUsed(rootClosure: rootClosure, 
+                                                 convertedAndReabstractedClosure: cvt.fromFunction,
+                                                 convertedAndReabstractedClosures: &convertedAndReabstractedClosures)
+  }
+
+  if let cvt = convertedAndReabstractedClosure as? ConvertEscapeToNoEscapeInst {
+    convertedAndReabstractedClosures.insert(cvt)
+    return 
+      markConvertedAndReabstractedClosuresAsUsed(rootClosure: rootClosure, 
+                                                 convertedAndReabstractedClosure: cvt.fromFunction,
+                                                 convertedAndReabstractedClosures: &convertedAndReabstractedClosures)
+  }
+
+  if let mdi = convertedAndReabstractedClosure as? MarkDependenceInst {
+    convertedAndReabstractedClosures.insert(mdi)
+    return 
+      markConvertedAndReabstractedClosuresAsUsed(rootClosure: rootClosure, convertedAndReabstractedClosure: mdi.value,
+                                                 convertedAndReabstractedClosures: &convertedAndReabstractedClosures)
+  }
+
+  fatalError("Unexpected instruction used during closure conversion/reabstraction: \(convertedAndReabstractedClosure)")
+}
+
+// ===================== Types ===================== //
+
+/// Represents all the information required to represent a closure in isolation, i.e., outside of a callsite context
+/// where the closure may be getting passed as an argument.
+///
+/// Composed with other information inside a `ClosureArgDescriptor` to represent a closure as an argument at a callsite.
+private struct ClosureInfo {
+  let closure: SingleValueInstruction
+  let lifetimeFrontier: [Instruction]
+
+  init(closure: SingleValueInstruction, lifetimeFrontier: [Instruction]) {
+    self.closure = closure
+    self.lifetimeFrontier = lifetimeFrontier
+  }
+
+}
+
+/// Represents a closure as an argument at a callsite.
+private struct ClosureArgDescriptor {
+  let closureInfo: ClosureInfo
+  /// The index of the closure in the callsite's argument list.
+  let closureArgumentIndex: Int
+  let parameterInfo: ParameterInfo
+  /// This is only needed if we have guaranteed parameters. In most cases
+  /// it will have only one element, a return inst.
+  var reachableExitBBs: [BasicBlock]
+}
+
+/// An intermediate data structure that holds the information about a closure at a callsite that will be specialized on
+/// it.
+///
+/// This type is an implementation detail and exists only to delay the creation of the final `ClosureArgDescriptor`s
+/// until the complete lifetime frontier of the corresponding closure has been computed.
+private struct IntermediateClosureArgDescriptor {
```

Comment:
The `ClosureArgDescriptor` contains a `ClosureInfo`. It's the `ClosureInfo` type that contains a `lifetimeFrontier`.

I wanted to have a separate `ClosureInfo` type because it represents a closure in isolation, i.e., outside of a call site context. And it composes with other information inside `ClosureArgDescriptor` to represent a closure as an argument at a call site.

The approach you're suggesting sounds similar enough to the current code. Is it mostly the name of the type that you don't like? I can simply change that to a named tuple if it seemed confusing?

Reply:
okay, fine with me